### DISCUSSION TITLE
Seeking Guidance on Efficient Data Fetching and Shared State between React Server Components in Next.js

### DISCUSSION THREAD

**Discussion by remibeaufils** [vercel/next.js#59885] [Help]
### Summary

Hello Next.js community!

I am currently facing a situation in my Next.js project and would appreciate some guidance or suggestions on how to handle it more efficiently.

I have this setup:
- `/app/community/layout.tsx`
- `/app/community/page.tsx`
- `/app/community/map/page.tsx`

All three components are implemented as React Server Components.

In the common layout `/community/layout.tsx`:
- there is a client component containing Links that facilitate navigation between the Community page and the Community Map page:

```tsx
const searchParams = useSearchParams();
<Link href={`/community${searchParams.size ? `?${searchParams.toString()}` : ""}`}>
  Community
</Link>
<Link href={`/community/map${searchParams.size ? `?${searchParams.toString()}` : ""}`}>
  Community Map
</Link>
```

- there is another client component containing dropdowns to navigate to the same page with searchParams updated:

```tsx
"use client";
export default function Filter({ placeholder, id, options }: Props) {
  const pathname = usePathname();
  <Dropdown
    options={options}
    defaultValue={searchParams.get(id) ?? undefined}
    onSelect={(value) => router.push(`${pathname}?${id}=${encodeURIComponent(value)}`)}
  />
}
```

Both pages use [searchParams](https://nextjs.org/docs/app/api-reference/file-conventions/page#searchparams-optional) to fetch data from a database. It's worth noting that both pages use the same function to fetch the data so given same searchParams they get the same data from the database; they just have different UI representations.

Additional context: the data is fetched from Supabase.

Now, when the website is already loaded, and I navigate from one page to the other, I want to avoid repeating the data-fetching process from the database. Indeed only the path in the url changes, not the searchParams and so the data-fetching process produces the same result. Instead, I'd like to find a way for the page to use the data already fetched by the other so the navigation would be quicker. I would also like to maintain the page as a React Server Component so I can't use a context to share this data.

My question is: How could I achieve this efficiently? Are there any recommended patterns or strategies for sharing fetched data between pages that are React Server Components and avoiding redundant data fetching during navigation within a section of the website?

From what I see, the only solution would be to use a cache. The cache would be used only when passing from a page to the other with the same searchParams and invalidate each time the searchParams change.

I would greatly appreciate any insights, code examples, or suggestions from the community.

Thank you in advance for your time and assistance.

### Example

https://patio.wow.art/community

**Comment by ananuness** [comment]
Hii! As you said, you can use cache and take advantage of the fact that next offers this functionality in native fetch. A good way for you to take advantage of this is to use [Route handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) to make requests, [configure the cache](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#caching-data), also, you can see this video about [revalidation and cache](https://www.youtube.com/watch?v=ztswJg7MYCs&t) to use in your server components. And there, you can [set cookies](https://nextjs.org/docs/app/api-reference/functions/cookies#cookiessetname-value-options) (if the data is not large) to make your work easier as well. Oh, and about sharing data between server components, you also can use fetch with cache or react's cache function for that, without worrying about making duplicate requests for the same data

> Important to know [when data is not cached](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching).

Now, considering that you are using Supabase, you can also check out this article about [Fetching and caching Supabase data in Next.js 13 Server Components](https://supabase.com/blog/fetching-and-caching-supabase-data-in-next-js-server-components). 

Finally, if you can use cookies, here we go with an example of a Route handler setting cookies
```js
export async function POST(req) {
  try {
    const body = await req.json();

    const user = await prisma.user.findUnique({
      where: { email: body.email },
    });

    const JWT_EXPIRES_IN = proccess.env.JWT_EXPIRES_IN";

    const token = await signJWT(
      { sub: user.id },
      { exp: `${JWT_EXPIRES_IN}m` }
    );

    const tokenMaxAge = parseInt(JWT_EXPIRES_IN) * 60;

    const cookieOptions = {
      name: "token",
      value: token,
      httpOnly: true,
      path: "/",
      secure: process.env.NODE_ENV !== "development",
      maxAge: tokenMaxAge,
    };

    const response = new NextResponse(
      JSON.stringify({
        status: "success",
        token,
      }),
      {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }
    );

    await Promise.all([
      response.cookies.set(cookieOptions),
      response.cookies.set({
        name: "logged-in",
        value: "true",
        maxAge: tokenMaxAge,
      }),
    ]);

    return response;
  } catch (error) {
    return NextResponse.json({ error: 'Internal Server Error' }, { status: 500 })
  }
}
```

I hope it helps and if you have any questions, don't hesitate to ask!
